compiler/src/dotty/tools/dotc/transform/patmat/Space.scala

```diff
@@ -355,18 +355,21 @@ class SpaceEngine(using Context) extends SpaceLogic {
   def project(pat: Tree): Space = pat match {
     case Literal(c) =>
       if (c.value.isInstanceOf[Symbol])
-        Typ(c.value.asInstanceOf[Symbol].termRef, false)
+        Typ(c.value.asInstanceOf[Symbol].termRef, decomposed = false)
       else
-        Typ(ConstantType(c), false)
+        Typ(ConstantType(c), decomposed = false)
 
     case pat: Ident if isBackquoted(pat) =>
-      Typ(pat.tpe, false)
+      Typ(pat.tpe, decomposed = false)
 
     case Ident(nme.WILDCARD) =>
-      Or(Typ(erase(pat.tpe.stripAnnots), false) :: constantNullSpace :: Nil)
+      if pat.tpe <:< defn.AnyValType then
```

Comment:
could we use [isNullableClass](https://github.com/lampepfl/dotty/blob/d2dd0832fec8416b934d730a7e785d79f6218777/compiler/src/dotty/tools/dotc/core/SymDenotations.scala#L800) instead? That way null won't pollute our patterns when -Yexplicit-nulls is enabled.

Reply:
also, for exhaustiveness we don't report missing nulls, so maybe it's fine to do the same for unreachability?
```scala
scala> final case class Foo(x: Int)
// defined case class Foo

scala> def foo(x: Foo) = x match { case Foo(_) => }
def foo(x: Foo): Unit

scala> foo(null)
scala.MatchError: null
```

Reply:
actually I guess we should emit a warning `_ only matches null` like we already do for `case _ =>`

Reply:
We have a feature that reports a warning if the last wildcard only matches `null` (https://github.com/lampepfl/dotty/pull/4253#discussion_r179486080). I think removing the `null` related stuff will make the code simpler.

Reply:
> actually I guess we should emit a warning `_ only matches null` like we already do for `case _ =>`

That would make the check more complex and expensive, while not very useful to users. I think removing `null` related logic will produce expected and helpful warnings to users while keeping the logic simple and fast.

Reply:
Actually, we can avoid computing precisely whether a pattern covers exactly `null`. Instead, we can blindly provide a hint for `null` if the last case is `_` and is unreachable. This allows us to use `Typ(tp)` instead of `Or(Type(tp), nullSpace)` for a wildcard pattern.

Doing that simplifies the code, and uncovers a bug that I'm now investigating:

```Scala
object Test {
  sealed trait Cause[+E]

  object Cause {
    final case class Fail[E](value: E) extends Cause[E]
  }

  def fn(cause: Cause[Any]): String =
    cause match {
      case Cause.Fail(t: Throwable) => t.toString
      case Cause.Fail(any)          => any.toString
    }
}
```

Previously no warning is produced because the checker thought that 2nd case matches `Fail(null)`, which is incorrect.